### ISSUE TITLE
Cypress can't connect to Firefox 100 inside a Docker container

### ISSUE THREAD

**Issue by kaiyoma** [cypress-io/cypress#22231]
### Current behavior

```
12:15:38  ====================================================================================================
12:15:38  
12:15:38    (Run Starting)
12:15:38  
12:15:38    ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
12:15:38    │ Cypress:        9.6.1                                                                          │
12:15:38    │ Browser:        Firefox 100 (headless)                                                         │
12:15:38    │ Node Version:   v16.13.0 (/usr/local/bin/node)                                                 │

...

12:16:30  Still waiting to connect to Firefox, retrying in 1 second (attempt 59/62)
12:16:30  Still waiting to connect to Firefox, retrying in 1 second (attempt 60/62)
12:16:31  Still waiting to connect to Firefox, retrying in 1 second (attempt 61/62)
12:16:32  Still waiting to connect to Firefox, retrying in 1 second (attempt 62/62)
12:16:32  Cypress failed to make a connection to Firefox.
12:16:32  
12:16:32  This usually indicates there was a problem opening the Firefox browser.
```

### Desired behavior

Cypress should work.

### Test code to reproduce

Run any tests against headless Firefox 100 in a Docker Linux environment.

### Cypress Version

9.6.1

**Comment by fabian-froehlich** [comment]
We had similar issues running cypress 9.7, while chrome had no issues.
```
Still waiting to connect to Firefox, retrying in 1 second (attempt 62/62)
Cypress failed to make a connection to Firefox.
This usually indicates there was a problem opening the Firefox browser.
Error: connect ECONNREFUSED 127.0.0.1:46333
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1142:16)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by kaiyoma** [comment]
Chrome has also been very reliable for us, but Firefox testing has been a huge PITA for the year or two I've been trying to make it work for us. We can barely get Cypress to connect to Firefox anymore, and in the rare event it does work, tests might fail for odd reasons, and video recording doesn't work. It's a pretty sorry state of affairs considering Cypress and Firefox are both mature products.

**Comment by AtofStryker** [comment]
> Chrome has also been very reliable for us, but Firefox testing has been a huge PITA for the year or two I've been trying to make it work for us. We can barely get Cypress to connect to Firefox anymore, and in the rare event it does work, tests might fail for odd reasons, and video recording doesn't work. It's a pretty sorry state of affairs considering Cypress and Firefox are both mature products.

Hey @kaiyoma. Thank you for opening an issue and I am sorry to hear about the issues you are running into with Cypress and Firefox. We do have an issue open for Firefox video recording, which is tracked in #18415. 

Currently, our prebuilts only contain [Firefox 99](https://github.com/cypress-io/cypress-docker-images/tree/master/browsers). We should have prebuilts for Firefox 100 and 101, but currently do not. In the meantime,  would you be able to share more details about the `Dockerfile` you are working with? I also put together a quick reproduction repository [here](https://github.com/AtofStryker/cypress-issue-22231) that I am more than happy to help fill out with details to reproduce this issue. Or if you'd like to clone and help fill out some missing details that works too!

**Comment by AtofStryker** [comment]
> We had similar issues running cypress 9.7, while chrome had no issues. `Image: cypress/browsers:node16.14.2-slim-chrome100-ff99-edge`
> 
> ```
> Still waiting to connect to Firefox, retrying in 1 second (attempt 62/62)
> Cypress failed to make a connection to Firefox.
> This usually indicates there was a problem opening the Firefox browser.
> Error: connect ECONNREFUSED 127.0.0.1:46333
>     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1142:16)
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> ```
> 
> With the migration to cypress 10.0.3 we disabled video capturing, removed a hook that listend on redux-event-bus and removed the cypress-terminal-report plugin. Firefox test is now able to start, but still crashes at some point:
> 
> ```
> Timed out waiting for the browser to connect. Retrying...
> Timed out waiting for the browser to connect. Retrying again...
> The browser never connected. Something is wrong. The tests cannot run. Aborting...
> The browser never connected. Something is wrong. The tests cannot run. Aborting...
>   (Results)
>   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
>   │ Tests:        0                                                                                │
>   │ Passing:      0                                                                                │
>   │ Failing:      1                                                                                │
>   │ Pending:      0                                                                                │
>   │ Skipped:      0                                                                                │
>   │ Screenshots:  0                                                                                │
>   │ Video:        false                                                                            │
>   │ Duration:     0 seconds                                                                        │
>   │ Spec Ran:     static_pages_test.cy.ts                                                          │
>   └────────────────────────────────────────────────────────────────────────────────────────────────┘
> Could not process 'close:browser:tabs'. No automation clients connected.
> Error: Could not process 'close:browser:tabs'. No automation clients connected.
>     at SocketE2E.onAutomation (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/socket-base.js:87:15)
>     at onAutomationClientRequestCallback (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/socket-base.js:133:25)
>     at /builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:76:20
>     at Promise.cancellationExecute [as _execute] (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/debuggability.js:406:9)
>     at Promise._resolveFromExecutor (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:518:18)
>     at new Promise (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:103:10)
>     at Automation.requestAutomationResponse (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:61:16)
>     at /builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:57:52
>     at /builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:111:28
>     at tryCatcher (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
>     at Function.Promise.attempt.Promise.try (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/method.js:39:29)
>     at Automation.normalize (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:89:38)
>     at /builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/lib/automation/automation.js:142:25
>     at tryCatcher (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/util.js:16:23)
>     at Promise._settlePromiseFromHandler (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:547:31)
>     at Promise._settlePromise (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:604:18)
>     at Promise._settlePromiseCtx (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/promise.js:641:10)
>     at _drainQueueStep (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:97:12)
>     at _drainQueue (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:86:9)
>     at Async._drainQueues (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:102:5)
>     at Immediate.Async.drainQueues [as _onImmediate] (/builds/project/cache/Cypress/10.0.3/Cypress/resources/app/packages/server/node_modules/bluebird/js/release/async.js:15:14)
>     at processImmediate (node:internal/timers:464:21)
> ```

@fabian-froehlich would you be able to share more details about your `Dockerfile`? It might help me understand how the image is being consumed in your use case.

**Comment by fabian-froehlich** [comment]
We are calling "yarn cypress run --browser firefox" using "node16.14.2-slim-chrome100-ff99-edge" as image in gitlab-ci. I could share details to the cypress config tomorrow. 

Are there any recommended settings for the docker host?

Edge and Chrome do not show such behavior.

**Comment by AtofStryker** [comment]
> We are calling `yarn cypress run --browser firefox` using `node16.14.2-slim-chrome100-ff99-edge` as image in gitlab-ci.
> 
> We are not building our own image, so there is no docker file.
> 
> Are there any recommended settings for the docker host?
> 
> I could share details to the cypress config tomorrow.
> 
> Edge and Chrome do not show such behavior.

Is this configured in a `gitlab-ci.yml` file? Would you be able to share the details of that? It would help me understand the context of how the image is run, as well as arguments passed into it.

**Comment by fabian-froehlich** [comment]
`.gitlab-ci.yml`
```
Cache:
  stage: cache
  cache:
    key: $CI_COMMIT_REF_SLUG
    paths:
      - cache/
      - node_modules/
      - yarn.lock.sha256sum
    policy: pull-push
  script:
    - yarn install --frozen-lockfile --prefer-offline
    - |
      YARN_LOCK_SHA256_HASH=$( sha256sum yarn.lock | awk '{ print $1 }')
      echo "Current sha256 hash is $YARN_LOCK_SHA256_HASH"
      echo "Checking sha256 hash of yarn.lock"

      if ! sha256sum -c yarn.lock.sha256sum; then
        echo "yarn.lock checksum does not match cache"
        sha256sum yarn.lock > yarn.lock.sha256sum
      else
        echo "Cache is the same as before, removing cache files to prevent cache from updating"
        rm -rf node_modules
        rm -rf .yarn
        rm -rf yarn.lock.sha256sum
        rm -rf cache/
      fi

Cypress:
  image: pvog-images.artifactory-extern.dataport.de/cypress/browsers:node16.14.2-slim-chrome100-ff99-edge
  stage: test
  needs: ["Cache", "Bundle App", "Unit Tests"]
  cache:
    key: ${CI_COMMIT_REF_SLUG}
    paths:
      - cache
      - node_modules
  parallel:
    matrix:
      - BROWSER:
          - chrome
          - firefox
          - edge
  script:
    # start Application under Test
    - yarn run start-with-url &
    # start Cypress
    - yarn run cy:run --browser $BROWSER
  timeout: 30 minutes
  artifacts:
    when: always
    paths:
      - cypress/videos/**/*.mp4
      - cypress/screenshots/**/*.png
      - cypress/results/**/*
      - cypress/logs
    reports:
      junit: ["cypress/results/*.xml"]
    expire_in: 1 day
  except:
    - master
    - tags
```


`package.json`
```
  "scripts": {
    "start-with-url": "webpack serve --config webpack.config.develop.js --port=3001 --env backend_url=http://127.0.0.1:1080",
    "cy:run": "cypress run",
...
  }
```

**Comment by AtofStryker** [comment]
created https://github.com/cypress-io/cypress-docker-images/issues/680 to update browser images for firefox 100 and 101. I will be trying to reproduce this issue shortly!

**Comment by AtofStryker** [comment]
Hey @fabian-froehlich. Sorry for the delay. I was able to create a quick reproduction [here](https://gitlab.com/AtofStryker/cypress-issue-22231/activity) within gitlab CI and firefox seems to be connecting as expected. The `gitlab.yml` file is a lot simpler I will say, but would you be able to clone/fork this and see if you get similar results, or are able to recreate the error?

**Comment by fabian-froehlich** [comment]
Hello @AtofStryker,
thanks for your reply and example project. I pushed it into our on-premise GitLab and did one tweak. The pipeline starts two jobs for Firefox and chrome in parallel. I scheduled the pipeline for every 30 minutes and had ~40 runs so far. Among these runes there is a failure-rate of 5% for Firefox and 0% for Chrome.
This pipeline is scheduled every 30 minutes and after ~40 runs I see two failures among Firefox and none failures among chrome. 5 % failure rate with Firefox.

Log: Failure-Example:
```
$ npm ci
added 169 packages, and audited 170 packages in 34s
27 packages are looking for funding
  run `npm fund` for details
found 0 vulnerabilities
$ npx cypress run --browser $BROWSER
[STARTED] Task without title.
[SUCCESS] Task without title.
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[596:0619/000157.400743:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[596:0619/000157.403673:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
====================================================================================================
  (Run Starting)
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        10.1.0                                                                         │
  │ Browser:        Firefox 99 (headless)                                                          │
  │ Node Version:   v16.14.2 (/usr/local/bin/node)                                                 │
  │ Specs:          1 found (spec.cy.ts)                                                           │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  spec.cy.ts                                                                      (1 of 1)
Timed out waiting for the browser to connect. Retrying...
Still waiting to connect to Firefox, retrying in 1 second (attempt 18/62)
...
Still waiting to connect to Firefox, retrying in 1 second (attempt 62/62)
Cypress failed to make a connection to Firefox.
This usually indicates there was a problem opening the Firefox browser.
Error: connect ECONNREFUSED 127.0.0.1:46073
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1161:16)
```
Log Success Example:
```
$ npx cypress run --browser $BROWSER
[STARTED] Task without title.
[SUCCESS] Task without title.
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[596:0619/083244.651645:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[596:0619/083244.655717:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
====================================================================================================
  (Run Starting)
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        10.1.0                                                                         │
  │ Browser:        Firefox 99 (headless)                                                          │
  │ Node Version:   v16.14.2 (/usr/local/bin/node)                                                 │
  │ Specs:          1 found (spec.cy.ts)                                                           │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  spec.cy.ts                                                                      (1 of 1)
  empty spec
    ✓ passes ([42](https://git.dataport.de/pvog/spielwiesen/cypress-debug/-/jobs/970647#L42)3ms)
  1 passing (2s)
  (Results)
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        1                                                                                │
  │ Passing:      1                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     2 seconds                                                                        │
  │ Spec Ran:     spec.cy.ts                                                                       │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
====================================================================================================
  (Run Finished)
       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  spec.cy.ts                               00:02        1        1        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        00:02        1        1        -        -        -
```

**Comment by AtofStryker** [comment]
@fabian-froehlich Thank you for providing this. To reproduce this, what all did you need to add to the `gitlab-ci.yml` file? could you PR the changes into the [reproduction](https://github.com/AtofStryker/cypress-issue-22231) and I can set up cron scheduling to see if I can reproduce it?

**Comment by fabian-froehlich** [comment]
@AtofStryker : There are in fact a lot of environment variables to setup proxy, no_proxy, etc/hosts, ...
But I don't see how this would make a job flaky. What happens if you schedule your pipeline every 30 Minutes for a couple of days? Is it stable?

The gitlab-ci.yml did not change much:
```
stages:
  - test

cache:
  key: ${CI_COMMIT_REF_SLUG}
  paths:
    - node_modules/
    - .npm/

test:
  image: cypress/browsers:node16.14.2-slim-chrome100-ff99-edge
  stage: test
  parallel:
    matrix:
      - BROWSER:
          - chrome
          - firefox
  script:
    - npm ci
    - npx cypress run --browser $BROWSER
  artifacts:
    when: always
    paths:
      - cypress/videos/**/*.mp4
      - cypress/screenshots/**/*.png
    expire_in: 1 day
```

**Comment by rsmartins78** [comment]
I have a similar situation. 
We use Github Actions as our CI engine, and with Cypress 8.3 + Firefox it works fine, but after updating Cypress to 10.2.0 and Firefox 101, cypress can't connect to firefox.
Locally, this combination works, it just fails on the CI environment.

Any clue?

**Comment by kaiyoma** [comment]
> > Chrome has also been very reliable for us, but Firefox testing has been a huge PITA for the year or two I've been trying to make it work for us. We can barely get Cypress to connect to Firefox anymore, and in the rare event it does work, tests might fail for odd reasons, and video recording doesn't work. It's a pretty sorry state of affairs considering Cypress and Firefox are both mature products.
> 
> Hey @kaiyoma. Thank you for opening an issue and I am sorry to hear about the issues you are running into with Cypress and Firefox. We do have an issue open for Firefox video recording, which is tracked in #18415.
> 
> Currently, our prebuilts only contain [Firefox 99](https://github.com/cypress-io/cypress-docker-images/tree/master/browsers). We should have prebuilts for Firefox 100 and 101, but currently do not. In the meantime, would you be able to share more details about the `Dockerfile` you are working with? I also put together a quick reproduction repository [here](https://github.com/AtofStryker/cypress-issue-22231) that I am more than happy to help fill out with details to reproduce this issue. Or if you'd like to clone and help fill out some missing details that works too!

@AtofStryker Here is our Dockerfile for e2e testing:

```
FROM cypress/base:16.13.0

SHELL ["/bin/bash", "-o", "pipefail", "-c"]

# Install Chrome
ARG CHROME_VERSION=102.0.5005.61-1
RUN curl -fsSL https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
  && echo "deb http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list \
  && apt-get update \
  && apt-get install -y --no-install-recommends google-chrome-stable=$CHROME_VERSION \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*

# Install Firefox
RUN curl -fsSL -o firefox.tar.bz2 https://download-installer.cdn.mozilla.net/pub/firefox/releases/100.0.2/linux-x86_64/en-US/firefox-100.0.2.tar.bz2 \
  && echo "e7b42e9fd82bcd569cbca1fc95b7e1a364c3dbf4d44a0624b4466ec1bd67f836  firefox.tar.bz2" > firefox.sha256 \
  && sha256sum -c firefox.sha256 \
  && tar -C /opt -xjf firefox.tar.bz2 \
  && rm firefox.sha256 firefox.tar.bz2 \
  && ln -fs /opt/firefox/firefox /usr/bin/firefox

# Upgrade npm
ARG NPM_VERSION=8.9.0
RUN npm install -g npm@$NPM_VERSION
```

**Comment by emilyrohrbough** [comment]
@kaiyoma Thank you for providing your Dockerfile. Unfortunately, I was unable to produce the issue. I was able to build the image and run my specs (used 1 and then 2) against Firefox without any issue. Can you provide debug logs with `DEBUG=cypress*` (no `:` to capture the verbose browser connection logs). I observed (likely) this same issue locally a few months ago on my Mac when running locally. I wasn't able to reproduce reliably, however what had happened was the Firefox instance had essentially hanged/never closed & Cypress has associated it "launched" so it would attempt to connect to the existing instance and would timeout.

If Firefox is throwing some sort of error and/or is unable to connect, we should be killing and restarting the instance (at least once IMO) to try to "recover" from this type of behavior. There is a parent issue I recently opened that seems to align with this (#22631).

I would like to review the verbose debug logs before grouping with the issue I linked and ensure there isn't something else wonky happening, but seems to be similar.

**Comment by mjhenkes** [comment]
Right now there doesn't seem to be enough information to reproduce the problem on our end. We'll have to close this issue until we can reproduce it. This does not mean that your issue is not happening - it just means that we do not have a path to move forward.

Please open a new issue with a reproducible example and link to this issue. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.
